Closes #1339 - Dataframe index set incorrectly on logical indexing

Ethan-DeBandi99 opened the pull request:
Close #1339 

- Updates index handling to identify `pdarrays` with `dtype=bool`. In this case, the index key is updated to `key=arange(key.size)[key]`.
- Updates index handling to identify `lists` of `bools`. In this case, the index key is updated `rows = arange(len(key))[key]` and then processed the same as an integer list.
- Adds testing for boolean indexing to ensure proper functionality

Ethan-DeBandi99 commented on arkouda/dataframe.py:
```diff
@@ -276,6 +277,14 @@ def __getitem__(self, key):
                     result._columns.append(k)
                 result._empty = False
                 return result
+            elif type(key[0]) == bool:
```
I don't have a problem with updating it to not allow lists of `int` or `bool`. I was assuming that eventually we would expand this to mimic `pandas.DataFrame` (per the discussion with users) to allow for more than just string column names and integer indexes. I can certainly pull this out for now though.

reuster986 replied:
Yeah, I think for now let's pull them out. Eventually, we can allow python lists of int/bool to index columns (but never rows).

mhmerrill replied:
Are we dropping another issue/PR for this change?

Ethan-DeBandi99 replied:
I just realized I put my comment on the wrong PR. Sorry I was using the GitHub app on my phone. I don't think we need a separate PR. I was planning to knock this out first thing Monday.

Comment I meant to put here and put in the wrong PR:
@reuster986 not going to lie, I am a bit confused as to why a list can't index into rows. Can you give me some details on your thought process here? I will get it pulled out though.

Ethan-DeBandi99 replied:
@reuster986 - I removed the `int` and `bool` cases from `__getitem__()` if a list is passed in. I am still not sure I fully understand why we don't want to allow a list to index to rows. It seems that would be handling it the same as a pdarray, which I think makes sense. Maybe there is a reason I am missing, but I do think we should reconsider allowing lists to index as they were. For the time being, I added an error message if a list is any type other than `str`.

reuster986 replied:
I could be convinced otherwise, but my reasoning is twofold:
1. Arkouda does not allow python lists to index pdarrays/Strings/Categorical, and indexing rows of DataFrames is essentially the same operation. If we allow python lists to index DataFrame rows, we should allow them to index pdarray, etc. But this makes me uncomfortable, because it is a mixed operation between client and server objects that implicitly transfers data from client to server. I'm a fan of forcing the user to be explicit about when and why to transfer data, because I've noticed that having some sense of where the data resides is crucial to writing performant arkouda code.
2. DataFrame indexing is inherently a little ambiguous about whether the target is rows or columns, and keeping the rule "python lists are for columns; pdarrays are for rows" seems like an intuitive way to delineate the two. I think most users have the intuition that python lists are for small objects, whereas pdarrays are for large objects, which matches up nicely with typical DataFrame dimensions (billions of rows by a few dozen columns).

joshmarshall1 reviewed (approved):
Looks good

Ethan-DeBandi99 commented:
@reuster986 - I removed the `int` and `bool` cases from `__getitem__()` if a list is passed in. I am still not sure I fully understand why we don't want to allow a list to index to rows. It seems that would be handling it the same as a pdarray, which I think makes sense. Maybe there is a reason I am missing, but I do think we should reconsider allowing lists to index as they were. For the time being, I added an error message if a list is any type other than `str`.

Ethan-DeBandi99 commented:
@reuster986 - That makes sense to me. Thanks for the explanation.
